Add the ability to actually use OpenSSL

mathias-lang-sociomantic opened the pull request:
Before, it was not possible to enable OpenSSL support in libdrizzle-redux since the configure script didn't support enabling it (though it required openssl to be installed).

Expected results (and performed tests, on linux):
- `./configure`
- `./configure --enable-ssl[=yes]`
- `./configure --disable-ssl` (or `./configure --enable-ssl=no`)

Followed by `ldd libdrizzle/.libs/libdrizzle-redux.so`

In the first and second case, I get:
```
	linux-vdso.so.1 =>  (0x00007fffc69e3000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f235b7f1000)
	libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f235b592000)
	libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f235b1b6000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f235aea4000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f235ab9e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f235a7d9000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f235a5c2000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f235a3be000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f235bc25000)
```

Before this P.R., and with the third version, I get:
```
	linux-vdso.so.1 =>  (0x00007ffecefbd000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f51f80c1000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f51f7daf000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f51f7aa9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f51f76e4000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f51f74cd000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f51f84f4000)
```

mathias-lang-sociomantic committed:
Add missing import for OpenSSL support in drizzle.cc

mathias-lang-sociomantic committed:
OpenSSL: Fix broken dependency configure.ac script (add ssl feature)

Previously, the code was checking that OpenSSL was present, and aborting compilation otherwise, but not actually doing anything.
The code expected the macro `USE_OPENSSL` to be defined, which wasn't, and the binary wasn't linked with `-lssl` either.

The configure script already provide a `--with-openssl=<path>` option, but it doesn't allow disabling the feature.
Changing it to be `--with-openssl=<no>|<path>[|yes]` would make it quite weird to use, so instead we separate enabling the *feature* of using OpenSSL to the actual dependency.
This could also be extended to use another crypto library in the future.

andreas-bok-sociomantic commented on configure.ac:
```diff
@@ -21,6 +21,32 @@ AC_CANONICAL_BUILD
 AC_ARG_PROGRAM
 AC_USE_SYSTEM_EXTENSIONS
 
+# SSL support, for encrypted connections
```
It should go into a separate macro and be included from the m4 dir.

andreas-bok-sociomantic replied:
It seems out of line with the format of the `configure.ac` to have it inline

andreas-bok-sociomantic commented:
Checked the build failure on travis 

```sh
libdrizzle/conn.cc:1261:80: error: implicit conversion loses integer precision:
      'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  ...SSL_read(con->ssl, (char*)con->buffer_ptr + con->buffer_size, available_buffer);
     ~~~~~~~~                                                      ^~~~~~~~~~~~~~~~
libdrizzle/conn.cc:1384:61: error: implicit conversion loses integer precision:
      'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
      write_size= SSL_write(con->ssl, con->buffer_ptr, con->buffer_size);
```
I have a docker image set up for cross compilation of the project. So I will fix that error

andreas-bok-sociomantic commented:
Maybe rephrase second ci msg to
 > OpenSSL: Fix broken enable/disable option in configure script

mathias-lang-sociomantic committed:
Only read up to INT_MAX data from the SSL connection

Because clang (rightfully) complains we are feeding a `size_t` instead of an `int`.
Not doing an actual cast because, heh, underflow.

mathias-lang-sociomantic commented:
Appeased the Travis gods. Gonna split the rest out on Monday :)

mathias-lang-sociomantic commented:
Updated

andreas-bok-sociomantic commented:
LGTM
